applications/nrf_desktop/configuration/common/ble_adv_def.h

```diff
@@ -56,7 +71,74 @@ static const struct bt_data ad_bonded[] = {
 	),
 };
 
-/* When using BT_LE_ADV_OPT_USE_NAME, device name is added to scan response
- * data by controller.
- */
-static const struct bt_data sd[] = {};
+#if CONFIG_DESKTOP_BLE_ADV_SWIFT_PAIR
+/* Ensure that grace period timeout is long enough for Swift Pair. */
+BUILD_ASSERT(CONFIG_CAF_BLE_ADV_GRACE_PERIOD_TIMEOUT >= 30);
+#endif
+
+
+static int fast_pair_adv_prepare(struct bt_data *adv_data, bool fp_discoverable)
+{
+	/* Make sure that Fast Pair data was freed and set to NULL to prevent memory leaks. */
+	if (adv_data->data) {
+		k_free((void *)adv_data->data);
+		adv_data->data = NULL;
+	}
+
+	/* Fast Pair pairing mode must be set manually. */
+	bt_fast_pair_set_pairing_mode(fp_discoverable);
+
+	size_t buf_size = bt_fast_pair_adv_data_size(fp_discoverable);
+	uint8_t *buf = k_malloc(buf_size);
+
+	if (!buf) {
+		return -ENOMEM;
+	}
+
+	int err = bt_fast_pair_adv_data_fill(adv_data, buf, buf_size, fp_discoverable);
+
+	if (err) {
+		k_free(buf);
+	}
+
+	return err;
+}
+
+static int prepare_adv_data(struct caf_ble_adv_state *state, struct caf_ble_adv_data *data)
```

Comment:
I am against adding such complex code inside header

Reply:
Ok, I will modify our approach regarding this one. Maybe we could remove def file and allow user to provide the function in another way.

I think allowing user to provide a function to prepare the advertising data could allow to define the data in very flexible way (e.g. data update could also depend on used local identity, number of bonds, grace period status or any other application specific parameters). Then whole advertising state machine will remain common (as it's part of CAF ble_adv module), but user still will be able to configure his undirected advertising data in a very flexible way.

Using advertising data providers requires maintaining states of providers which may not be best approach in some use-cases. For example user may want to use various advertising data per local identity or benefit from extended advertising. The providers API may introduce some problems when it comes to triggering adv data update too (after using provider API, user must remember to trigger update of actually broadcasted data). At some point application must use provider API to decide about advertising data so user will need to provide this part of the implementation anyway.